### PR TITLE
Added instrumentation to Hive copy.

### PR DESCRIPTION
**Pull request by ibuenros**
Adds timers to Hive distcp work unit generation.
Also adds a `MultiTimingEvent` class that emits a single event for timings of different stages in a process.

### CODE REVIEW

**Comment by ibuenros** [comment]
@abti can you review?

**Comment by ibuenros** [comment]
@zliu41 can you review?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 1]
```diff
@@ -199,6 +220,10 @@
 
     log.info("Finding copy entities for table " + dataset.table.getCompleteName());
 
+    this.eventSubmitter = new EventSubmitter.Builder(dataset.getMetricContext(), "hive.dataset.copy").build();
+    Closer closer = Closer.create();
```
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: closer init should be in try block and closer.closer() in finally to be on the safer side

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 2]
```diff
@@ -434,17 +463,22 @@ private Partition getTargetPartition(Partition originPartition, Path targetLocat
     private final Partition partition;
     private final Properties properties;
     private Optional<Partition> existingTargetPartition;
+    private final EventSubmitter eventSubmitter;
 
     public PartitionCopy(Partition partition, Properties properties) {
       this.partition = partition;
       this.properties = properties;
       this.existingTargetPartition = Optional.fromNullable(targetPartitions.get(this.partition.getValues()));
+      this.eventSubmitter = new EventSubmitter.Builder(dataset.getMetricContext(), "dataset.hive.copy").
```
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Second param to EventSubmitter.Builder is inconsistent with HiveEntityCopyHelper; there the second argument was "hive.dataset.copy, here it is "dataset.hive.copy"

File: gobblin-metrics/src/main/java/gobblin/metrics/event/MultiTimingEvent.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  /**
+   * Ends the current stage and submits the event containing the timings of each event.
+   * @param additionalMetadata additional metadata to include in the event.
+   * @throws IOException
+   */
+  public void submit(Map<String, String> additionalMetadata) throws IOException {
+
+    if (this.submitted) {
```
Comment (abti) [inline gobblin-metrics/src/main/java/gobblin/metrics/event/MultiTimingEvent.java]: MultiTimingEvent seems thread unsafe during close(), can we call it out in java docs?

**Comment by abti** [comment]
@ibuenros .. a few comments, rest looks good to me

**Comment by ibuenros** [comment]
Fixed. Merging.

**Pull request merged**